### ISSUE TITLE
Is Mac OS supported？

### ISSUE THREAD

**Issue by Jzow** [guillaume-be/rust-bert#209]
I'm at REDME.md only sees the manual installation of windows and Linux

**Comment by guillaume-be** [comment]
Hello @Jzow ,

Yes - MacOS is supported, a build is part of the [continuous integration tests](https://github.com/guillaume-be/rust-bert/runs/4635320194?check_suite_focus=true). 

CPU should work with the pre-built libtorch binaries at https://pytorch.org/get-started/locally/ (or would get downloaded automatically when building the library). Please note that as far as I know, M1 chips are however not yet supported.

**Comment by Jzow** [comment]
@guillaume-be thanks I try see

**Comment by recurrence** [comment]
I was able to build and run a CPU m1 version using the apple silicon torch binaries.

**Comment by raphaelmenges** [comment]
`rust-bert` works for me as well on my M1 Mac when (1) installing libtorch via homebrew `brew install libtorch` and (2) adding `export LIBTORCH=/opt/homebrew/Cellar/libtorch/1.10.1_1` and `export DYLD_LIBRARY_PATH=${LIBTORCH}/lib:$DYLD_LIBRARY_PATH` to my `.zprofile`.

@recurrence: How did you acquire the libtorch version for M1/ARM64? I could not find a public release outside of homebrew.

**Comment by recurrence** [comment]
@raphaelmenges  I pulled the arm64 libs from the pip3 torch install.

**Comment by winston0410** [comment]
> `rust-bert` works for me as well on my M1 Mac when (1) installing libtorch via homebrew `brew install libtorch` and (2) adding `export LIBTORCH=/opt/homebrew/Cellar/libtorch/1.10.1_1` and `export DYLD_LIBRARY_PATH=${LIBTORCH}/lib:$DYLD_LIBRARY_PATH` to my `.zprofile`.
> 
> @recurrence: How did you acquire the libtorch version for M1/ARM64? I could not find a public release outside of homebrew.

I have done that, and tried using `conda install pytorch` and link the library with `LD_LIBRARY_PATH`, but I am hitting the same error. @recurrence Have you get that previously? I am using 1.11.0 of Pytorch

```
 cargo:rerun-if-env-changed=TORCH_CUDA_VERSION
  cargo:rerun-if-env-changed=LIBTORCH
  cargo:rustc-link-search=native=/opt/homebrew/Cellar/libtorch/1.11.0/lib
  cargo:rerun-if-changed=libtch/torch_api.cpp
  cargo:rerun-if-changed=libtch/torch_api.h
  cargo:rerun-if-changed=libtch/torch_api_generated.cpp.h
  cargo:rerun-if-changed=libtch/torch_api_generated.h
  cargo:rerun-if-changed=libtch/stb_image_write.h
  cargo:rerun-if-changed=libtch/stb_image_resize.h
  cargo:rerun-if-changed=libtch/stb_image.h
  cargo:rerun-if-env-changed=LIBTORCH_CXX11_ABI
  TARGET = Some("aarch64-apple-darwin")
  OPT_LEVEL = Some("3")
  HOST = Some("aarch64-apple-darwin")
  CXX_aarch64-apple-darwin = None
  CXX_aarch64_apple_darwin = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_aarch64-apple-darwin = None
  CXXFLAGS_aarch64_apple_darwin = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("aes,crc,dit,dotprod,dpb,dpb2,fcma,fhm,flagm,fp16,frintts,jsconv,llvm14-builtins-abi,lor,lse,neon,paca,pacg,pan,pmuv3,ras,rcpc,rcpc2,rdm,sb,sha2,sha3,ssbs,v8.1a,v8.2a,v8.3a,v8.4a,vh")
  running: "c++" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-arch" "arm64" "-I" "/opt/homebrew/Cellar/libtorch/1.11.0/include" "-I" "/opt/homebrew/Cellar/libtorch/1.11.0/include/torch/csrc/api/include" "-Wl,-rpath=/opt/homebrew/Cellar/libtorch/1.11.0/lib" "-std=c++14" "-D_GLIBCXX_USE_CXX11_ABI=1" "-o" "/Users/hugosum/Desktop/jyutping-tools/target/debug/build/torch-sys-218041f83084dd22/out/libtch/torch_api.o" "-c" "libtch/torch_api.cpp"
  cargo:warning=clang: warning: -Wl,-rpath=/opt/homebrew/Cellar/libtorch/1.11.0/lib: 'linker' input unused [-Wunused-command-line-argument]
  cargo:warning=libtch/torch_api.cpp:953:9: error: no member named 'tracer' in namespace 'torch::jit'; did you mean 'torch::tracer'?
  cargo:warning=    if (torch::jit::tracer::isTracing())
  cargo:warning=        ^~~~~~~~~~~~~~~~~~
  cargo:warning=        torch::tracer
  cargo:warning=libtch/torch_api.h:16:5: note: expanded from macro 'PROTECT'
  cargo:warning=    x \
  cargo:warning=    ^
  cargo:warning=/opt/homebrew/Cellar/libtorch/1.11.0/include/ATen/TracerMode.h:113:11: note: 'torch::tracer' declared here
  cargo:warning=namespace tracer {
  cargo:warning=          ^
  cargo:warning=libtch/torch_api.cpp:953:29: error: no member named 'isTracing' in namespace 'at::tracer'
  cargo:warning=    if (torch::jit::tracer::isTracing())
  cargo:warning=                    ~~~~~~~~^

```

**Comment by recurrence** [comment]
Oh bummer, perhaps some changes have happened since.  I recently reformatted this machine for a new client.  I'll take some time tonight to try to get this up and running again and report back :)

**Comment by recurrence** [comment]
Hey peeps, sorry this took so long. Work got crazy during the week.  So on this relatively clean M1Max I did this (at first I accidentally ran it in a conda which was configured for x86 and was confused when the torch libs were the wrong arch :P )

```
pip3 install torch
pip3 show # Just to see where torch installed
cp -R /opt/homebrew/lib/python3.9/site-packages/torch .

export LIBTORCH=/<PATH-TO-RUST-BERT-FOLDER>/torch
export LD_LIBRARY_PATH=${LIBTORCH}/lib:$LD_LIBRARY_PATH

cargo build

cp torch/lib/* ~/lib/ 
```

I need to double check why it was looking for them in ~/lib after building

cargo build and cargo build —release both succeed

I then ran the summarization example and it dumped out 
Downloading https://cdn.huggingface.co/sshleifer/distilbart-cnn-6-6/vocab.json [877.76KiB]... ✓ Done! Finished in 0 seconds
Downloading https://cdn.huggingface.co/sshleifer/distilbart-cnn-6-6/merges.txt [445.62KiB]... ✓ Done! Finished in 0 seconds
Downloading https://cdn.huggingface.co/sshleifer/distilbart-cnn-6-6/config.json [1.58KiB]... ✓ Done! Finished in 0 seconds
Downloading https://huggingface.co/sshleifer/distilbart-cnn-6-6/resolve/main/rust_model.ot [635.06MiB]......... ✓ Done! Finished in 11 seconds
 K2-18b is not too hot and not too cold for liquid water to exist. This is the first such discovery in a planet in its star's habitable zone. The presence of water vapour was confirmed in the atmosphere of K2, a planet circling a star in the constellation Leo.
